File: sdks/typescript/src/wire/http-fetch-proto-wire.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025 PlexSpaces Contributors
//
// Manual protobuf wire for plexspaces.wasm.v1 HttpFetchRequest / HttpFetchResponse.
// Matches sdks/go/plexspaces/http_fetch_proto_wire.go.

import { appendLengthDelimited, readLengthDelimited, readVarint, skipField } from './proto-wire-common.js';

function utf8Valid(bytes: Uint8Array): boolean {
  try {
    new TextDecoder('utf-8', { fatal: true }).decode(bytes);
    return true;
  } catch {
    return false;
  }
}

function bytesToBase64Sync(bytes: Uint8Array): string {
  let bin = '';
  for (let i = 0; i < bytes.length; i++) bin += String.fromCharCode(bytes[i]!);
  if (typeof btoa !== 'undefined') return btoa(bin);
  const Buf = (globalThis as unknown as { Buffer?: { from: (b: Uint8Array) => { toString: (enc: string) => string } } }).Buffer;
  if (Buf) return Buf.from(bytes).toString('base64');
  throw new Error('base64 encode unavailable');
}

/** Build HttpFetchRequest: repeated map entry (field 1), body (field 2). */
export function encodeHttpFetchRequestWire(headers: Record<string, string>, body: Uint8Array | null): Uint8Array {
  let buf = new Uint8Array(0);
  const enc = new TextEncoder();
  for (const [k, v] of Object.entries(headers)) {
    const kb = new Uint8Array(enc.encode(k));
    const vb = new Uint8Array(enc.encode(v));
    let entry = appendLengthDelimited(new Uint8Array(0), 1, kb);
    entry = appendLengthDelimited(entry, 2, vb);
    buf = appendLengthDelimited(buf, 1, entry);
  }
  const bodyUse = body && body.length > 0 ? new Uint8Array(body) : new Uint8Array(0);
  buf = appendLengthDelimited(buf, 2, bodyUse);
  return buf;
}

function parseStringStringMapEntry(entry: Uint8Array): { key: string; val: string } {
  let pos = 0;
  let key = '';
  let val = '';
  while (pos < entry.length) {
    const { value: tag, n: tn } = readVarint(entry, pos);
    pos += tn;
    const fn = Number(tag >> 3n);
    const wt = Number(tag & 7n);
    if (fn === 1 && wt === 2) {
      const { slice, nextPos } = readLengthDelimited(entry, pos);
      pos = nextPos;
      key = new TextDecoder('utf-8', { fatal: false }).decode(slice);
    } else if (fn === 2 && wt === 2) {
      const { slice, nextPos } = readLengthDelimited(entry, pos);
      pos = nextPos;
      val = new TextDecoder('utf-8', { fatal: false }).decode(slice);
    } else {
      pos = skipField(entry, pos, wt);
    }
  }
  return { key, val };
}

export function decodeHttpFetchResponseWire(data: Uint8Array): {
  status: number;
  headers: Record<string, string>;
  body: string;
} {
  const out = {
    status: 0,
    headers: {} as Record<string, string>,
    body: '',
  };
  let pos = 0;
  while (pos < data.length) {
    const { value: tag, n: tn } = readVarint(data, pos);
    pos += tn;
    const fn = Number(tag >> 3n);
    const wt = Number(tag & 7n);
    if (fn === 1 && wt === 0) {
      const { value: v, n: m } = readVarint(data, pos);
      pos += m;
      out.status = Number(v);
    } else if (fn === 2 && wt === 2) {
      const { slice: sl, nextPos } = readLengthDelimited(data, pos);
      pos = nextPos;
      const { key, val } = parseStringStringMapEntry(sl);
      if (key) out.headers[key] = val;
    } else if (fn === 3 && wt === 2) {
      const { slice: sl, nextPos } = readLengthDelimited(data, pos);
      pos = nextPos;
      out.body = utf8Valid(sl) ? new TextDecoder('utf-8', { fatal: false }).decode(sl) : bytesToBase64Sync(sl);
    } else {
      pos = skipField(data, pos, wt);
    }
  }
  return out;
}

File: sdks/go/plexspaces/http_fetch_proto_wire.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025 PlexSpaces Contributors
//
// Manual protobuf wire for plexspaces.wasm.v1.HttpFetchRequest / HttpFetchResponse.
// Used by TinyGo WASM http-fetch (WIT: link-name, method, path-and-query, request: payload).
// Matches prost encoding in crates/wasm-runtime simple_component_host.

package plexspaces

import (
	"encoding/base64"
	"fmt"
	"unicode/utf8"
)

// encodeHttpFetchRequestWire builds HttpFetchRequest: map headers (field 2), body (field 3).
func encodeHttpFetchRequestWire(headers map[string]string, body []byte) ([]byte, error) {
	var buf []byte
	for k, v := range headers {
		if !utf8.ValidString(k) || !utf8.ValidString(v) {
			return nil, fmt.Errorf("http fetch request header key/value must be UTF-8")
		}
		var entry []byte
		entry = appendLengthDelimited(entry, 1, []byte(k))
		entry = appendLengthDelimited(entry, 2, []byte(v))
		buf = appendLengthDelimited(buf, 2, entry)
	}
	if body == nil {
		body = []byte{}
	}
	buf = appendLengthDelimited(buf, 3, body)
	return buf, nil
}

// decodeHttpFetchResponseWire parses HttpFetchResponse into the map shape Host.HTTPFetch returns.
func decodeHttpFetchResponseWire(data []byte) (map[string]any, error) {
	out := map[string]any{
		"status":  float64(0),
		"headers": map[string]any{},
		"body":    "",
	}
	headers := map[string]any{}
	pos := 0
	for pos < len(data) {
		tag, n, err := readVarint(data, pos)
		if err != nil {
			return nil, err
		}
		pos += n
		fn := int(tag >> 3)
		wt := tag & 7
		switch {
		case fn == 2 && wt == 0:
			v, m, err := readVarint(data, pos)
			if err != nil {
				return nil, err
			}
			pos += m
			out["status"] = float64(v)
		case fn == 3 && wt == 2:
			sl, np, err := readLengthDelimited(data, pos)
			if err != nil {
				return nil, err
			}
			pos = np
			k, v, err := parseStringStringMapEntry(sl)
			if err != nil {
				return nil, err
			}
			headers[k] = v
		case fn == 4 && wt == 2:
			sl, np, err := readLengthDelimited(data, pos)
			if err != nil {
				return nil, err
			}
			pos = np
			if utf8.Valid(sl) {
				out["body"] = string(sl)
			} else {
				out["body"] = base64.StdEncoding.EncodeToString(sl)
			}
		default:
			pos, err = skipField(data, pos, wt)
			if err != nil {
				return nil, err
			}
		}
	}
	out["headers"] = headers
	return out, nil
}

func readLengthDelimited(data []byte, pos int) ([]byte, int, error) {
	ln, n, err := readVarint(data, pos)
	if err != nil {
		return nil, pos, err
	}
	pos += n
	end := pos + int(ln)
	if end > len(data) {
		return nil, pos, fmt.Errorf("length-delimited field truncated")
	}
	return data[pos:end], end, nil
}

func parseStringStringMapEntry(entry []byte) (key, val string, err error) {
	pos := 0
	for pos < len(entry) {
		tag, n, err := readVarint(entry, pos)
		if err != nil {
			return "", "", err
		}
		pos += n
		fn := int(tag >> 3)
		wt := tag & 7
		switch {
		case fn == 1 && wt == 2:
			sl, np, err := readLengthDelimited(entry, pos)
			if err != nil {
				return "", "", err
			}
			pos = np
			key = string(sl)
		case fn == 2 && wt == 2:
			sl, np, err := readLengthDelimited(entry, pos)
			if err != nil {
				return "", "", err
			}
			pos = np
			val = string(sl)
		default:
			pos, err = skipField(entry, pos, wt)
			if err != nil {
				return "", "", err
			}
		}
	}
	return key, val, nil
}
